### PR TITLE
Strsplittrail

### LINKED ISSUE #26: Align `col_strsplit` to `strsplit`, warts and all?
Consider:

    > strsplit(" a b ", " ")
    [[1]]
    [1] ""  "a" "b"

    > col_strsplit(" a b ", " ")
    [[1]]
    [1] ""  "a" "b" ""

I'd argue that the `crayon` output is more correct.  It is, however, different from what `strsplit` does.  Would you consider accepting a PR that "fixes" this?  If no, not a big deal.  I'm going to have to work around this issue anyway since for the time being I'm relying on the CRAN version of this package.

### PR DESCRIPTION
**Pull request by brodieG**
Proposed "fix" for #26 (Fixes #26).  A few noteworthy comments:

* I "fixed" this issue only in `col_strsplit`.  It could potentially be "fixed" in `crayon:::re_table` which would be cleaner, but would affect multiple other functions
* I also enabled the `empty` parameter for `non_matching`; since it wasn't being used (I think), I updated all calls to that function to specify `empty=TRUE` to avoid changing behavior

Let me know if you want me to change / revert any of this.

### CODE REVIEW

**Comment by brodieG** [comment]
Fyi, I'll submit the resolved conflicts later today.

**Comment by brodieG** [comment]
@gaborcsardi, fyi, the branch is now up to date.

**Comment by brodieG** [comment]
@gaborcsardi, let me know if you intend to eventually merge this one or not.  Fine if you don't, but it would be helpful for me to know one way or other since it affects how I write my package that imports crayon.

File: R/string_operations.r [hunk 1]
```diff
@@ -212,6 +214,14 @@ col_strsplit <- function(x, split, ...) {
   plain <- strip_style(x)
   splits <- re_table(split, plain, ...)
   chunks <- non_matching(splits, plain, empty = TRUE)
+  # drop any zero length trailing chunks that are created when the split
+  # happens at end of string; this is to replicate 'substr' behavior
+  chunks <- lapply(
+    chunks,
+    function(y)
+      if(nrow(y) > 1L && !tail(y, 1L)[, "length"]) head(y, -1L) else y
```
Comment (gaborcsardi) [inline R/string_operations.r]: Why the `nrow(y) > 1L` condition? I have 
```r
❯ strsplit("", "a")[[1]]
character(0)
```
so we want to drop a single `""` as well, I think.

Comment (brodieG) [inline R/string_operations.r]: I'll have to look at it again later.  Clearly this is wrong, but I think I was trying to ensure that we don't drop this:

```
> strsplit("a", "a")[[1]]
[1] ""
```
Hopefully I can look at it this evening.

**Comment by brodieG** [comment]
Okay, so unfortunately addressing the corner case you highlight is not completely straightforward.  This latest push is *NOT* ready for merging as there remains some corner cases and I want to test this more thoroughly.

I put it here for your inspection in because the (currently partial) solution I put in is not particularly elegant.  As far as I can tell, I can't easily rely on the existing infrastructure to handle the corner cases, as passing 0-row chunks causes `col_substr` and co. to break.  Instead, I have to handle the corner cases in `col_strsplit` directly.  I could probably fix the internal functions to be more elegant, but that will likely mean modifying a fair bit more code.

Let me know if have any concerns with me just handling the corner cases as the `col_strsplit`.

**Comment by gaborcsardi** [comment]
Thanks! I'll try to look at it this evening.

**Comment by brodieG** [comment]
Okay, this should be good to go now, assuming you're okay with how I've addressed the issues.  I wish `strsplit` was a bit more consistent in how it handles corner cases.

File: tests/testthat/test-operations.R [hunk 2]
```diff
@@ -119,8 +128,30 @@ test_that("col_strsplit multiple strings", {
...
+  # Longer strings
+  expect_identical(
+    col_strsplit(c("", "a", "aa"), "a"), strsplit(c("", "a", "aa"), "a")
+  )
+  expect_identical(
+    col_strsplit(c("abaa", "ababza"), "b."), strsplit(c("abaa", "ababza"), "b.")
+  )
+})
 
-  expect_equal(col_strsplit("", "-"), list(""))
```
Comment (brodieG) [inline tests/testthat/test-operations.R]: Note that these tests no longer behave like they used to so I changed them.

**Comment by brodieG** [comment]
I addressed the last remaining issue I opened.  Note that one issue is that `substr("abc", "hello", "hello")` works (returns NA), whereas I just sidestepped that by disallowing string position arguments in `col_substr`.

Also, appveyor is failing, but I think the issue is not related to the code changes I made (for some reason `devtools` didn't load properly).

Finally, haven't heard much one way or another on this pull request.  Feedback would be appreciated.  Thanks.

**Comment by gaborcsardi** [comment]
I am sorry this is taking long. I'll look at it today.

**Comment by brodieG** [comment]
Let me know if you want me to simplify this pull request by breaking it up into smaller ones.

**Comment by gaborcsardi** [comment]
Thanks much! I am sorry that it took so long!

**Comment by brodieG** [comment]
Great, thanks for accepting the pull.
